### DISCUSSION TITLE
Warning message for undefined defaultValue

### DISCUSSION THREAD

**Discussion by wenchonglee** [react-hook-form/react-hook-form#2797] [General]
Hey,

In #2717, a warning message was added to `controller.tsx` for missing `defaultValue`. 

For me, there has been a flood of warning messages since I updated react-hook-form because the components I am using accepts `undefined` as "empty". 
To expand on my use-case, sometimes I only populate a default value if a preference setting is true:

```
const defaultValues = {
    user: useCurrentUser ? currentUser : undefined
}
const methods = useForm({defaultValues})
```

Can we find a way to mute such a warning message?

**Comment by bluebill1049** [comment]
defaultValue shouldn't be `undefined`, perhaps use empty string?

**Reply by wenchonglee** [reply]
I am using the `controller` in conjunction with a component that uses object values, [this in particular](https://elastic.github.io/eui/#/forms/combo-box).
`undefined` means no option is selected

**Reply by bluebill1049** [reply]
`undefined` doesn't work well with `Controller`. perhaps `{}` or `null`?

**Reply by wenchonglee** [reply]
Hmm, I was thinking more along the lines of letting people intentionally have `undefined` without causing warning messages. 

I'll try to see if I can work around it, thanks for the quick response as always. :)

**Reply by bluebill1049** [reply]
np

**Comment by darron1217** [comment]
I also intentionally use undefined as defaultValue. Are there any other workarounds?

**Reply by RLRabinowitz** [reply]
Same here. We use a different form library and wanna switch over to `react-hook-form`, but `undefined` not being supported as a value is really problematic for us. It seems that `setValue` of `undefined` sets the value to its `defaultValue`, which is not expected behaviour.

This basically means that when I submit, our backend needs to accept a different default for optional fields, either `null`s or `''`s, instead of `undefined`s. And I don't backend behaviour because of this. @bluebill1049 What's the library's proposed solution for this issue?

**Comment by andrwo** [comment]
I also find this quite a troublesome issue.

It will be much friendlier to use if the library does not impose a requirement that all fields must have a default, and just default the initial value to 'undefined' as the default if the node is not found in the defaultValue tree.

**Comment by wenchonglee** [comment]
Since this discussion has been gaining some traction, I thought I'd unmark the answer.

I've personally been trying to avoid the warning messages by using null but I've found it more of a chore than anything. 
I also agree with what @andrwo said about not imposing such a requirement from the library's standpoint

**Comment by alexandrzavalii** [comment]
any updates on this issue? undefined still seams to cause problems

**Comment by fabien-roy** [comment]
Still quite a problem.

We shouldn't force forms to have valid values at first render. It makes sense to begin the form in an invalid state, then let the user enter valid values.

For TypeScript users, at least we don't get a type error. For instance, I have a component to handle a form, which receives `defaultValues` as a prop. Using `Partial<T>` for this prop, then passing it to react-hook-form works as intended, so at least there's that.
